### ISSUE TITLE
StaticPointers?

### ISSUE THREAD

**Issue by pkamenarsky** [purescript/purescript#2869]
[StaticPointers Haskell spec](https://downloads.haskell.org/~ghc/7.10.1/docs/html/users_guide/static-pointers.html).

For PureScript on the backend, one particular use case for static pointers would be straightforward client/server RPC.

Has this been discussed before?

**Comment by paf31** [comment]
Can you please provide a proposal for how this might be implemented?

**Comment by pkamenarsky** [comment]
Basically:

* Make `static` a reserved keyword. There's no need to extend the AST, since this is just basic function application.
* Define `data Static_Ptr a = ...` in e.g. `purescript-static`
* Define `data Static_Value = forall a. Static_Value a` in `purescript-static`
* After `desugarImports` walk through all `App "static" expr` nodes and replace those with `Constructor "Static_Ptr" unique_id`. What this does is replacing all `static expr` expressions with `Static_Ptr unique_id`
* Before replacing every `static` application, check if all free variables in said expression reference top-level definitions in the same or some other imported module; if not, throw an error
* Export a `static_ptr_table :: Map String Static_Value` map per module which maps unique_ids to expressions
* Define `unsafeDerefPointer :: Static_Ptr a -> a` in `purescript-static`

I'm not sure how to best implement `unsafeDerefPointer`. Is it possible to dereference a module-specific export at runtime? If yes, `unsafeDerefPointer` would just unpack the value containted in the module specific `static_ptr_table`. The module name could be stored in `Static_Ptr` itself. If not, all module specific `static_ptr_tables` will need to be merged at the end into a single big `static_ptr_table`, possibly in the packaging step.

**Comment by hdgarrood** [comment]
Could we do this in a library if we had some kind of compile-time metaprogramming, maybe like Template Haskell, that allows you to ask for the current module name and things like that?

**Comment by pkamenarsky** [comment]
If `AST -> AST` transformations were possible, I don't see a problem. But my understanding is that Template PureScript (or whatever it ends up being called) is far off yet, while static pointers should be a relatively straightforward change with the only implication that `static` must be a reserved keyword.

**Comment by pkamenarsky** [comment]
I took the liberty and implemented a quick [proof of concept](https://github.com/purescript/purescript/compare/master...pkamenarsky:static-pointers).

The following code, compiled with [purescript-static](https://github.com/pkamenarsky/purescript-static),

`Remote.purs`

    module Remote where

    import Static

    id_static :: forall a. StaticPtr (a -> a)
    id_static = static \x -> x

`Test.purs`

    module Test where

    import Control.Monad.Eff
    import Control.Monad.Eff.Console

    import Prelude
    import Static

    import Remote

    main :: forall eff. Eff (console :: CONSOLE | eff) Unit
    main = do
      log $ derefStaticPtr id_static "test_static"

prints out `test_static`.

`Remote.purs` generates the following code:

    // Generated by purs version 0.11.4
    "use strict";
    var Static = require("../Static");
    var static_ptr_table = {
        static_ptr_0: function (x) {
            return x;
        }
    };
    var id_static = new Static.StaticPtr("Remote", "static_ptr_0");
    module.exports = {
        static_ptr_table: static_ptr_table,·
        id_static: id_static
    };

Static pointers can't be defined in the same module in which they are dereferenced right now (hence the separate `Remote` module), because by the time `purescript-static/derefStaticPtrByModule` is called the exports table isn't set yet. Also, `derefStaticPtrByModule` uses the convention used in `psc-bundle` and looks up the static table in the global `PS` object. Don't know if e.g. Webpack differs here.

I don't think that issue is particularly hard to solve though. Please let me know if you like the general direction of this.

**Comment by paf31** [comment]
Thank you for writing this up. I'll think more about it soon.

**Comment by hdgarrood** [comment]
I'm not sure how I feel about the proposal but I do think it's unacceptable to refer to a global `PS` object; we don't make any assumption like that anywhere else in the compiler, and I'm pretty confident that many people's workflows depend on us not making that assumption.

**Comment by paf31** [comment]
Yes, we definitely can't rely on `PS`, since even that name is configurable. It should depend on the codegen backend, which might mean it needs core modifications.

I think it would be worth trying to find a way to implement this purely in a library too.

**Comment by pkamenarsky** [comment]
@paf31 re implementing this in a library, what do you have in mind?

**Comment by paf31** [comment]
I would explore how embedded DSLs could be used to solve the problem. You design a DSL for the problem you're trying to solve and then identify the closed terms. You can serialize the terms themselves for transport, or assign them unique IDs.

**Comment by pkamenarsky** [comment]
Maybe I'm misunderstanding, but isn't this just the standard manual way of doing distributed communication? Design a data type, serialize it, send it over the wire.

**Comment by paf31** [comment]
Yes, it is I suppose. So I guess I'm asking, is there enough of a benefit here that it's worth adding a whole new language feature?

I'm not really seeing how to build a safe abstraction on top of this, since it seems to rely crucially on an unsafe operation.

What is the advantage of the `static` approach over this sort of approach?

**Comment by pkamenarsky** [comment]
Actually it can be made safe; I glossed over some details in the proposal.

The main idea is that `static` has in fact the following signature:

    static :: forall a. Typeable a => a -> StaticPtr a

The `TypeRep` of `a` is then stored along the static expression in the static pointer table. One can then extract a `StaticKey` from a `StaticPtr`, which is just a wrapped unique identifier:

    staticKey :: StaticPtr a -> StaticKey

`StaticKey` is then sent over the wire. To look up a static expression on the receiving end, there is `derefStaticPtr`:

    derefStaticPtr :: forall a. Typeable a => StaticKey -> Maybe a

In that way, the unsafe operation is guarded behind a comparison of the `TypeRep` stored in the static pointer table with the `TypeRep` of the expected type. If the `TypeRep`s don't match, return Nothing.

In fact, static pointers could make RPC calls even safer. `StaticKey` could contain a fingerprint of the static expression AST, which would then be matched against the stored fingerprint in the static pointer table. This guarantees that a static expression can't be evaluated when the client expects it to be something different than what it actually is on the server, even if the types still match.

Even better, every time the server is deployed one could dump a json file containing all static expression fingerprints (triggered with a psc flag or whatever). Then, during development, one could always check whether the client code still references valid static expressions on the server. Or the compiler could issue warnings/errors on every stale static expression, externalizing the answer to the question "should I redeploy the server now?" into the compiler. It would also make for a very streamlined interactive editing experience -- change a function and have all now stale static expressions referencing that function light up in your editor.

On the devops side of things, one could even prevent deployments of clients with stale static expressions.

I don't think that level of end-to-end safety could even be possible with the traditional approach without any additional compiler support or some sort of preprocessing step that parses wire datatypes and computes fingerprints.

Since PureScript is starting to be a viable backend language, I speculate that streamlining the one-codebase-everywhere experience by removing tons of unnecessary RPC code is worth a relatively non-invasive language extension.

**Comment by TheMatten** [comment]
As I understand it, `StaticPointers` would be very useful for memoization in frameworks like Halogen - if we can assign static identity to some function and control it's inputs (by enforcing it being closed), we can easily compare such functions during patching by comparing their inputs and `StaticKey`-s. Halogen currently does some hacks using JS's `===` to achieve something similar, but the solution seems to be buggy and depends on user supplying properly constructed function.

**Comment by rhendric** [comment]
Am I correct that this could be done with a CoreFn transformation? If we get #3339, using custom CoreFn transformations becomes a much more practical option, and I think pretty obviously the preferable approach versus extending the language.
